[SPARK-32892][CORE][SQL] Fix hash functions on big-endian platforms.

mundaym opened the pull request:
MurmurHash3 and xxHash64 interpret sequences of bytes as integers
encoded in little-endian byte order. This requires a byte reversal
on big endian platforms.

I've left the hashInt and hashLong functions as-is for now. My
interpretation of these functions is that they perform the hash on
the integer value as if it were serialized in little-endian byte
order. Therefore no byte reversal is necessary.

### What changes were proposed in this pull request?
Modify hash functions to produce correct results on big-endian platforms.

### Why are the changes needed?
Hash functions produce incorrect results on big-endian platforms which, amongst other potential issues, causes test failures.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests run on the IBM Z (s390x) platform which uses a big-endian byte order.

mundaym committed:
[SPARK-32892][CORE][SQL] Fix hash functions on big-endian platforms.

MurmurHash3 and xxHash64 interpret sequences of bytes as integers
encoded in little-endian byte order. This requires a byte reversal
on big endian platforms.

I've left the hashInt and hashLong functions as-is for now. My
interpretation of these functions is that they perform the hash on
the integer value as if it were serialized in little-endian byte
order. Therefore no byte reversal is necessary.

This commit fixes existing tests on the IBM Z (s390x) platform.
I've modified the XXH64 tests to expect the same results on all
platforms. Other tests assume little-endian which aligns with the
specification for xxHash64.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

maropu commented:
ok to test

maropu commented:
cc: @rednaxelafx @srowen @kiszk

SparkQA commented:
**[Test build #128738 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128738/testReport)** for PR 29762 at commit [`df1a1c8`](https://github.com/apache/spark/commit/df1a1c8d826ec4d5cf64d8c04eb03e6dd5bb9d9d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33362/
Test PASSed.

SparkQA commented:
**[Test build #128738 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128738/testReport)** for PR 29762 at commit [`df1a1c8`](https://github.com/apache/spark/commit/df1a1c8d826ec4d5cf64d8c04eb03e6dd5bb9d9d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128738/
Test FAILed.

mundaym commented:
In an ideal world we'd probably consolidate platform endianness handling into the Platform class by adding methods that perform the necessary conversions. I'm not sure if that would be a reasonable to add to the Platform class API? We could, for example, have overloads for getters and setters for at least the primitive types that take a byte order argument so we could make calls like `Platform.getInt(base, offset, ByteOrder.LITTLE_ENDIAN)`. The implementation would be trivial:

```java
public static int getInt(Object object, long offset, ByteOrder order) {
  int x = _UNSAFE.getInt(object, offset);
  return ByteOrder.nativeOrder().equals(order) ? x : Integer.reverseBytes(x);
}
```

That would make changes like this one far cleaner. Though it might be slightly more work (and therefore be slightly riskier performance-wise) for the JIT compiler to optimize well than the proposed inline conversions in this PR.

maropu commented:
> We could, for example, have overloads for getters and setters for at least the primitive types that take a byte order argument so we could make calls like Platform.getInt(base, offset, ByteOrder.LITTLE_ENDIAN).

Yea, I think that approach can have performanec regression..

kiszk commented:
I also think that the proposed `getInt()` introduces the overhead.


The access primitives such as `getInt()` does not usually require byte reverse if the method is used within an primitive unit (e.g. Integer). This is the usual usages in Spark. This is because Spark (and Java) uses machine-native endian within an primirive unit.

IMHO, the usages of reverse order are the exceptional cases

mundaym commented:
Thanks, it sounds like the inline code that this PR uses is preferred to Platform class changes at this point so I'll leave this PR as-is.

It might be worth reassessing this later if/when `jdk.internal.misc.Unsafe` is adopted since that introduces `getIntUnaligned` and `getLongUnaligned` which take into account endianness and alignment requirements on the target platform. Using those functions would make this code fully portable and - as part of the standard library - should be well optimized by the compiler. Certainly the unaligned load and the reverse bytes are both treated as intrinsics by Hotspot. I think we'd need to wait until JDK 8 support is dropped to adopt it though.

srowen commented on common/sketch/src/main/java/org/apache/spark/util/sketch/Murmur3_x86_32.java:
```diff
@@ -92,8 +96,10 @@ private static int hashBytesByInt(Object base, long offset, int lengthInBytes, i
       int halfWord = Platform.getInt(base, offset + i);
-      int k1 = mixK1(halfWord);
-      h1 = mixH1(h1, k1);
+      if (isBigEndian) {
```
I am not sure how much difference it makes, but this condition is of course checked every time in the loop.
If one were willing to duplicate the loop, you could check it once at the outset. But unless there's evidence that is non-trivial overhead and that this is a very hot loop, probably the extra branch is negligible, esp. if it is smart enough to know it's always false/true.

dongjoon-hyun replied:
cc @rednaxelafx

HyukjinKwon replied:
also @kiszk

kiszk replied:
I agree to move this condition out of the loop while the overhead of the condition itself may be small.   
I expect that the HotSpot compiler would/could generate one native instruction `reverse load` with the adjacent getInt and reverseBytes.

mundaym replied:
I'm inclined to agree with @srowen that duplicating the loop is probably unnecessary. Hotspot will almost certainly be able to remove the branch and even if it doesn't this code isn't actually in the critical path for the hash function (the critical path is through mixH1) so the latency of the extra check can be hidden on an out of order processor. I'm happy to do it if the consensus is that it is warranted though.

srowen reviewed (approved):
This change seems OK as-is.

Another possible pattern is to make some object that encapsulates all endian-specific methods, and return a big- or little-endian implementation from a factory method. Then the method implementations can assume endianness and optimize, and don't need to check endianness. That said, it means duplicating code. So, probably for another time if it becomes important.

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #128780 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128780/testReport)** for PR 29762 at commit [`df1a1c8`](https://github.com/apache/spark/commit/df1a1c8d826ec4d5cf64d8c04eb03e6dd5bb9d9d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33403/
Test PASSed.

SparkQA commented:
**[Test build #128780 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128780/testReport)** for PR 29762 at commit [`df1a1c8`](https://github.com/apache/spark/commit/df1a1c8d826ec4d5cf64d8c04eb03e6dd5bb9d9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128780/
Test PASSed.

maropu commented:
> Another possible pattern is to make some object that encapsulates all endian-specific methods, and return a big- or little-endian implementation from a factory method.

Yea, that approach sounds resonable to me.

srowen reviewed (approved):
Any objections? seems fair enough.

kiszk commented on sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/XXH64Suite.java:
```diff
@@ -73,42 +72,22 @@ public void testKnownByteArrayInputs() {
+    Assert.assertEquals(0x0EAB543384F878ADL,
+            hasher.hashUnsafeBytes(BUFFER, Platform.BYTE_ARRAY_OFFSET, SIZE));
+    Assert.assertEquals(0xCAA65939306F1E21L,
+            XXH64.hashUnsafeBytes(BUFFER, Platform.BYTE_ARRAY_OFFSET, SIZE, PRIME));
```
Can we add test cases for `XXH64.hashUnsafeWords` and `hasher.XXH64.hashUnsafeWords`??

mundaym replied:
Done. I've tried to keep the style the same as the existing tests but I can refactor if desired.

kiszk commented:
I expect to provide additional tests. Regarding the condition, it is nice to have with me.

mundaym committed:
[SPARK-32892][TEST][FOLLOWUP] Add XXH64 hashUnsafeWords tests.

Add some additional test coverage for the hashUnsafeWords methods
provided by XXH64 as requested in the review of PR #29762.

mundaym commented:
> I expect to provide additional tests. Regarding the condition, it is nice to have with me.

Added some basic test coverage to hashUnsafeWords.

SparkQA commented:
**[Test build #128952 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128952/testReport)** for PR 29762 at commit [`cbfa4d6`](https://github.com/apache/spark/commit/cbfa4d63efba01fd06747260e252bac466d99853).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33574/
Test PASSed.

SparkQA commented:
**[Test build #128952 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128952/testReport)** for PR 29762 at commit [`cbfa4d6`](https://github.com/apache/spark/commit/cbfa4d63efba01fd06747260e252bac466d99853).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128952/
Test PASSed.

kiszk commented:
Thank you. The code and tests alone look good.

I have one question about use cases for `hashUnsafeWords`. Regarding `hashUnsafeBytes`. it make sense to return the same hash value against the same byte sequence. How about `hashUnsafeWords()`?   
In general, a word could have the different byte sequence on little-endian and big-endian (e.g. `78563412` v.s. `12345678`). While I see the caller of `hashUnsafeWords()`, I am not sure what hash value is expected.

mundaym commented:
> While I see the caller of hashUnsafeWords(), I am not sure what hash value is expected.

It's a good question. Right now the only users of `hashUnsafeWords` appear to be `UnsafeRow` and `BytesToBytesMap`. `UnsafeRow` uses 8 byte slots to store all primitive values. `BytesToBytesMap` appears to enforce that keys are 8 byte aligned. In both cases I am fairly certain they are calling `hashUnsafeWords` rather than `hashUnsafeBytes` simply as an optimization since they know the input is 8 byte aligned.

For `UnsafeRow` in particular sub-8-byte types do not appear to be extended to 8 byte types, just left aligned in the same slot. For example, a `float` would be left aligned in the slot rather than converted to a `double` to fill the slot. A word-by-word hash of them would therefore still produce different results on big- and little-endian systems.

My current understanding is therefore that `hashUnsafeWords` and `hashUnsafeBytes` should produce identical results.

Aside: the use of 'word' here is in my opinion ambiguous. Is it 32 bits on a 32 bit system? I think that if we wanted to be able to hash an array of 8 byte values as if they were encoded in little-endian byte order we should name the method after the Java type identifier, so something like `hashUnsafeLongs`.

kiszk commented:
I see. Make sense. Thank you very much.

srowen commented:
Merged to master
